### PR TITLE
Implement Input Mapping

### PR DESCRIPTION
**Pull request by zicklag**
Here's two players controlled both with gamepads, and keyboard controls work too!

punchy02-players

### CODE REVIEW

**Comment by zicklag** [comment]
@odecay this is ready for review now.

It replaces _all_ game input with mapable actions in three categories: `player`, `camera`, and `menu`. There is no longer any occurrence of `KeyCode` anywhere in the code, it's all mapped and loaded from the `GameMeta` asset. :)

You can also play two players, now, though I believe there's a weird bug where the second player cannot move while the first player is flopping. I haven't investigated yet, but I think we can fix that after this PR.

I'm going to follow up with a PR to allow you to navigate the menus with the gamepad, and then after that I'll look into getting two players working without issue.

**Comment by odecay** [comment]
> hough I believe there's a weird bug where the second player cannot move while the first player is flopping

Maybe its because of something I did in state checking for that system? Using return instead of break? https://github.com/fishfight/punchy/blob/0060a90bd2fed813a0bf5ae06593ec0b914a5ed1/src/main.rs#L664

**Comment by odecay** [comment]
> it replaces all game input with mapable actions in three categories: player, camera, and menu. There is no longer any occurrence of KeyCode anywhere in the code, it's all mapped and loaded from the GameMeta asset. :)

Awesome! Review + merge will be my first priority after the v0.0.2 release is up today.

> I'm going to follow up with a PR to allow you to navigate the menus with the gamepad, and then after that I'll look into getting two players working without issue.

:+1:  Sounds good!

File: assets/default.game.yaml [hunk 1]
```diff
@@ -12,6 +12,161 @@ main_menu:
     image: ui/main-menu-background.png
     image_size: [1919, 1027]
 
+default_input_maps:
```
Comment (odecay) [inline assets/default.game.yaml:15]: Very nice how clean it is to have analogs + dpad + keyboard all share input maps

File: Cargo.toml [hunk 2]
```diff
@@ -42,4 +43,7 @@ opt-level = 1
 lto = true
 codegen-units = 1 # Improves physics performance for release builds
 
-#TODO turn this into a debug feature
+[patch.crates-io]
+# Patch to enable improved DPad support. It's a merge of the following PR to Bevy 0.7.0 so we can
+# use it before 0.8 is release: https://github.com/bevyengine/bevy/pull/5220.
```
Comment (odecay) [inline Cargo.toml:48]: checked out the patch, looks relatively simple lets hope it gets merged :+1:

**Pull request merged**